Add optimizer from bayesian-optimization

spline2hg opened the pull request:
This PR adds the optimizer from `bayesian-optimization` library.

### Implemented Features
- [x] Basic Bayesian Optimization algorithm interface
- [x] Support for built-in acquisition functions (UCB, EI, POI)
- [x] Sequential Domain Reduction (SDR)
- [x] Added documentation and tests

### Remaining Work
- [ ] Support for nonlinear constraints  
  _constraint handling is currently skipped. Some changes are required in the way constraints are processed, before this can be properly implemented._
- [ ] Implementation of meta-acquisition functions (ConstantLiar and GPHedge)

spline2hg commented on docs/source/algorithms.md:
```diff
@@ -4043,6 +4043,79 @@ these optimizers, you need to have
+
+    bayes_opt supports the following options:
+
+    - **init_points** (int): Number of random exploration points to evaluate before 
+      starting optimization. Default is 5.
+
+    - **n_iter** (int): Number of Bayesian optimization iterations to perform after 
+      the initial random exploration. Default is 25.
```
Yes, these are the package defaults. Should we explicitly mention in the documentation that these values come from the package defaults? Also, to make them adaptive, how could we relate these parameters to the dimensionality of the problem?

janosg replied:
If they are the package defaults we don't need to deviate and don't make them adaptive. 

Yes, please mention that these are the package defaults.

spline2hg commented on src/optimagic/optimizers/bayesian_optimizer.py:
```diff
@@ -0,0 +1,280 @@
+class BayesOpt(Algorithm):
+    """Bayesian Optimization wrapper using bayes_opt package.
+
+    Args:
```
sure i will keep the one in algorithms.md

janosg commented on src/optimagic/optimizers/bayesian_optimizer.py:
```diff
@@ -0,0 +1,347 @@
+        Array of parameter values.
+
+    """
+    return np.array([kwargs[f"param{i}"] for i in range(n_params)])
```
Wouldn't `np.array(list(kwargs.values))` do the same without requiring a user to pass `n_params`?

Also: `kwargs` is not a good name here. Maybe use `params_dict` instead?

janosg commented on tests/optimagic/optimizers/test_bayesian_optimizer.py:
```diff
@@ -0,0 +1,145 @@
+    np.testing.assert_array_equal(result, np.array([1.0, 2.0]))
+
+
+def test_process_bounds():
+    """Test processing bounds for Bayesian optimization."""
+    bounds = InternalBounds(lower=np.array([-1.0, 0.0]), upper=np.array([1.0, 2.0]))
+    result = _process_bounds(bounds)
+    expected = {"param0": (-1.0, 1.0), "param1": (0.0, 2.0)}
+    assert result == expected
+
+    bounds_none = InternalBounds(lower=None, upper=np.array([1.0, 2.0]))
+    with pytest.raises(
+        ValueError, match="Bayesian optimization requires finite bounds"
+    ):
+        _process_bounds(bounds_none)
+
+    bounds_inf = InternalBounds(
+        lower=np.array([-1.0, 0.0]), upper=np.array([1.0, np.inf])
+    )
+    with pytest.raises(
+        ValueError, match="Bayesian optimization requires finite bounds"
+    ):
+        _process_bounds(bounds_inf)
```
Let's split this into three tests for the three cases so we get more information if any of them fails.

janosg reviewed (commented):
Looks very nice. Just two very small comments.

spline2hg commented:
@timmens, the pre-commit mypy check is passing, but the mypy in environment is raising an error about assigning to Any.Adding ignore to it is causing pre-commit mypy to complain.
Do we need to change the way we’re handling the optional dependency with bayes_opt?

janosg commented:
Have you tried to re-create your local conda environment to get the same version of mypy as in the CI runs?

timmens commented:
@spline2hg @janosg 

I've had a look at the situation and have a proposal. I did not, however, find out why this is a problem here and not with the `iminuit` library (my hunch is that it is because `bayes_opt` is typed but `iminuit` is not). Let me know what you think!

**Proposal:**

1. We import all runtime dependencies inside of functions that are only called when the optional dependency (here `bayes_opt`) is installed.
2. We import all required type annotations separately in a `TYPE_CHECKING` block.
3. We put the type annotations that come from the optional dependency in quotes so that the code does not fail at runtime if the optional dependency is not installed


**Example:**

### 1.

```python
def _solve_internal_problem(
    self, problem: InternalOptimizationProblem, x0: NDArray[np.float64]
) -> InternalOptimizeResult:
    if not IS_BAYESOPT_INSTALLED:
        raise NotInstalledError(
            "To use the 'bayes_opt' optimizer you need to install bayes_opt. "
            "Use 'pip install bayesian-optimization'. "
            "Check the documentation for more details: "
            "https://bayesian-optimization.github.io/BayesianOptimization/index.html"
        )
        
    from bayes_opt import BayesianOptimization, SequentialDomainReductionTransformer

```

### 2.

```python
if TYPE_CHECKING:
    from bayes_opt import BayesianOptimization
    from bayes_opt.acquisition import AcquisitionFunction
```

Since we import the runtime dependencies inside the function, we only need these objects for the type checking, and therefore also only if `TYPE_CHECKING` is true. If the optional dependency is not installed this would raise an ImportError, but we only type check in an environment where this is installed. If we want to avoid this altogether we could add `and IS_BAYES_OPT_INSTALLED`.


### 3.

```python
def _process_bayes_opt_result(
    optimizer: "BayesianOptimization",
    x0: NDArray[np.float64],
    problem: InternalOptimizationProblem,
) -> InternalOptimizeResult:
```

**Implications:**

If we decide to follow this route, then we should do the same for iminuit!

**Additionally:**

We should remove `bayes_opt` from `[[tool.mypy.overrides]]` since it is actually typed.

spline2hg commented:
@timmens Thanks for the detailed proposal! I had missed that bayes_opt is actually typed ,and that was causing the issue, thanks for pointing it out. 

I’ve added the following changes:
- runtime imports inside required functions
- type imports under TYPE_CHECKING
- added from __future__ import annotations so annotations are automatically stringified (instead explicit quotes)
- dropped the ignore_missing_imports stanza for bayes_opt

Happy to mirror the same fix for iminuit in the same PR.

janosg commented:
Hi @lesshaste, it's a perfectly fine question. We are super close to also finish the addition of nevergrad. I think we'll have a really amazing new release next week.
